### PR TITLE
 Add units sent by clicking a source and target stronghold

### LINKED ISSUE #15: Add units
Units should be able to move from one castle to another. Must have
  * attack power
  * speed
  * unit count
  * owner

### LINKED ISSUE #19: Add ability to send units
Clicking on a stronghold owned by the player, followed by clicking on another stronghold sends half the units to the target stronghold.

Cancel the initial stronghold selection if the second click was not on another stronghold.

### PR DESCRIPTION
**Pull request by salahusama**
Closes #15 
Closes #19 

#### Current functionality
1. Click on a stronghold _(this will be the source of the unit)_
2. Click on a stronghold _(this will be the target the unit goes to)_
3. Half the magnitude of the source stronghold will be sent
    * As a spherical unit added to the scene
4. Added or subtracted as appropriate on arrival to target
    * Player switches if needed _(color updated as well)_
5. Spherical unit removed from the scene

#### Notes
* Multiple units can be sent at the same time

### CODE REVIEW

**Comment by salahusama** [comment]
@QUzair, I had removed some of the strongholds for testing, but I do not remember the layout you had. It should not effect us for now though and we can have an issue for level design later.

File: Assets/GameEngine.cs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+                SendUnit(sourceInfo, selected);
+            }
+
+            source = null;
+        }
+    }
+
+    private void SendUnit(ObjectOwnerState sourceInfo, Transform selected) {
+        // Create unit at source stronghold
+        GameObject unit = Instantiate(unitPrefab, source.position, Quaternion.identity);
```
Comment (hashdoge) [inline Assets/GameEngine.cs]: What is Quaternion.identity?

File: Assets/GameEngine.cs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+        int strongholdMagnitude = sourceInfo.magnitude;
+        int unitMagnitude = strongholdMagnitude / 2;
```
Comment (hashdoge) [inline Assets/GameEngine.cs]: What do these two lines do?

File: Assets/GameEngine.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+        // Only send unit if it has a magnitude
+        if (unitMagnitude > 0) {
```
Comment (hashdoge) [inline Assets/GameEngine.cs]: What if there is a magnitude of 1? will it send half a unit?

File: Assets/GameEngine.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+            ObjectOwnerState unitInfo = unit.GetComponent<ObjectOwnerState>();
+
+            unitInfo.magnitude = unitMagnitude;
+            unitInfo.player = sourceInfo.player;
+            unit.GetComponent<UnitState>().SetSource(source.gameObject);
+            unit.GetComponent<Movement>().SetTarget(selected);
+
+            sourceInfo.magnitude -= unitMagnitude;
```
Comment (hashdoge) [inline Assets/GameEngine.cs]: Can you explain all of this code please

File: Assets/Movement.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
+﻿using UnityEngine;
+
+public class Movement : MonoBehaviour {
+
+    public float speed = 20f;
+
+    private Transform target;
+
+	void Start () {
+        if (target == null) {
```
Comment (hashdoge) [inline Assets/Movement.cs]: Why does it need to be null first?

File: Assets/UnitInteraction.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
+﻿using UnityEngine;
+
+public class UnitInteraction : MonoBehaviour {
+
+    private void OnTriggerEnter(Collider other) {
```
Comment (hashdoge) [inline Assets/UnitInteraction.cs]: explain this

File: Assets/UnitState.cs [hunk 7]
```diff
@@ -0,0 +1,14 @@
+﻿using UnityEngine;
+
+public class UnitState : MonoBehaviour {
```
Comment (hashdoge) [inline Assets/UnitState.cs]: what does this do?

File: Assets/GameEngine.cs [hunk 8]
```diff
@@ -0,0 +1,43 @@
+﻿using UnityEngine;
+
+public class GameEngine : MonoBehaviour {
+
+    public GameObject unitPrefab;
+    private Transform source;
+
+	public void RegisterStrongholdClick(Transform selected) {
```
[the remark this answers is not in the document]
Comment (salahusama) [reply Assets/GameEngine.cs]: I'll add extra comments for this, but what it essentially does is determine if the clicked stronghold is the source or target. It then calls `SendUnit()` which actually sends the units.

File: Assets/GameEngine.cs (hunk 1 above)
Comment (salahusama) [reply Assets/GameEngine.cs]: `Quaternion` is used for rotation. [`Quaternion.identity`](https://docs.unity3d.com/ScriptReference/Quaternion-identity.html) means no rotation.

File: Assets/GameEngine.cs (hunk 2 above)
Comment (salahusama) [reply Assets/GameEngine.cs]: Sets unit magnitude to half that of the source stronghold.

File: Assets/GameEngine.cs (hunk 3 above)
Comment (salahusama) [reply Assets/GameEngine.cs]: No because magnitude is an `int`, so 1 / 2 => 0

File: Assets/GameEngine.cs (hunk 4 above)
Comment (salahusama) [reply Assets/GameEngine.cs]: Yep. So we instantiated the Unit object above, but did not assign it an owner or a magnitude. We do so here. We calculated `unitMagnitude` a few lines up _(half the stronghold magnitude)_, the owner is the owner of the source stronghold as that's where they're leaving from.

We also set the source and the target of the unit. The target is where it moves to _(there is a script that updates the position of the unit)_. The source is used in collision detection to ignore any collisions with the source while leaving the stronghold.

Finally, we subtract the unit magnitude _(as the unit is leaving)_ from the source stronghold's magnitude.

File: Assets/UnitInteraction.cs (hunk 6 above)
Comment (salahusama) [reply Assets/UnitInteraction.cs]: `OnTrigger` is called when the unit enters a stronghold. The `Collider` in this case is the unit. I added a tag to all Units, and I check that to make sure the interaction is in fact a Unit interaction.

I get the source from the unit _(we set the source of the unit in the game engine above)_ and make sure this is not a collision caused by the unit colliding with the stronghold it came from which happens while the Unit is leaving the stronghold.

Then it calls `ComputeResult()` which updates the owner of this stronghold and the it's new magnitude.

File: Assets/UnitState.cs (hunk 7 above)
Comment (salahusama) [reply Assets/UnitState.cs]: This is not the best name to be fair, but it essentially just stores the source stronghold of the unit object. Used in the code above.

**Pull request merged**